client/SDL/sdl_disp.cpp

```diff
@@ -386,151 +297,132 @@ BOOL sdl_disp_handle_display_event(sdlDispContext* disp, const SDL_DisplayEvent*
-	sdlDispContext* sdlDisp;
-	rdpSettings* settings;
-
 	WINPR_ASSERT(disp);
 
-	sdlDisp = (sdlDispContext*)disp->custom;
-	WINPR_ASSERT(sdlDisp);
-	WINPR_ASSERT(sdlDisp->sdl);
+	auto sdlDisp = reinterpret_cast<sdlDispContext*>(disp->custom);
+	return sdlDisp->DisplayControlCaps(maxNumMonitors, maxMonitorAreaFactorA,
+	                                   maxMonitorAreaFactorB);
+}
 
-	settings = sdlDisp->sdl->common.context.settings;
+UINT sdlDispContext::DisplayControlCaps(UINT32 maxNumMonitors, UINT32 maxMonitorAreaFactorA,
+                                        UINT32 maxMonitorAreaFactorB)
+{
+	auto settings = _sdl->common.context.settings;
 	WINPR_ASSERT(settings);
 
 	WLog_DBG(TAG,
 	         "DisplayControlCapsPdu: MaxNumMonitors: %" PRIu32 " MaxMonitorAreaFactorA: %" PRIu32
 	         " MaxMonitorAreaFactorB: %" PRIu32 "",
 	         maxNumMonitors, maxMonitorAreaFactorA, maxMonitorAreaFactorB);
-	sdlDisp->activated = TRUE;
+	_activated = TRUE;
 
 	if (settings->Fullscreen)
 		return CHANNEL_RC_OK;
 
 	WLog_DBG(TAG, "DisplayControlCapsPdu: setting the window as resizable");
-	return sdl_disp_set_window_resizable(sdlDisp) ? CHANNEL_RC_OK : CHANNEL_RC_NO_MEMORY;
+	return set_window_resizable() ? CHANNEL_RC_OK : CHANNEL_RC_NO_MEMORY;
 }
 
-BOOL sdl_disp_init(sdlDispContext* sdlDisp, DispClientContext* disp)
+BOOL sdlDispContext::init(DispClientContext* disp)
 {
-	rdpSettings* settings;
-
-	if (!sdlDisp || !sdlDisp->sdl || !disp)
+	if (!disp)
 		return FALSE;
 
-	settings = sdlDisp->sdl->common.context.settings;
+	auto settings = _sdl->common.context.settings;
 
 	if (!settings)
 		return FALSE;
 
-	sdlDisp->disp = disp;
-	disp->custom = (void*)sdlDisp;
+	_disp = disp;
+	disp->custom = this;
 
 	if (settings->DynamicResolutionUpdate)
 	{
-		disp->DisplayControlCaps = sdl_DisplayControlCaps;
+		disp->DisplayControlCaps = sdlDispContext::DisplayControlCaps;
 	}
 
-	update_resizeable(sdlDisp->sdl, TRUE);
+	update_resizeable(_sdl, TRUE);
 	return TRUE;
 }
 
-BOOL sdl_disp_uninit(sdlDispContext* sdlDisp, DispClientContext* disp)
+BOOL sdlDispContext::uninit(DispClientContext* disp)
 {
-	if (!sdlDisp || !disp)
+	if (!disp)
 		return FALSE;
 
-	sdlDisp->disp = nullptr;
-	update_resizeable(sdlDisp->sdl, FALSE);
+	_disp = nullptr;
+	update_resizeable(_sdl, FALSE);
 	return TRUE;
 }
+
+sdlDispContext::sdlDispContext(sdlContext* sdl) : _sdl(sdl)
+{
+	WINPR_ASSERT(_sdl);
+	WINPR_ASSERT(_sdl->common.context.settings);
+	WINPR_ASSERT(_sdl->common.context.pubSub);
+
+	auto settings = _sdl->common.context.settings;
+	auto pubSub = _sdl->common.context.pubSub;
+
+	_lastSentWidth = _targetWidth = settings->DesktopWidth;
+	_lastSentHeight = _targetHeight = settings->DesktopHeight;
+	PubSub_SubscribeActivated(pubSub, sdlDispContext::OnActivated);
+	PubSub_SubscribeGraphicsReset(pubSub, sdlDispContext::OnGraphicsReset);
+	PubSub_SubscribeTimer(pubSub, sdlDispContext::OnTimer);
+}
+
+sdlDispContext::~sdlDispContext()
+{
+	if (_sdl)
```

Comment:
We already have an assertion for this in the body of the constructor, so this is unneeded (always true).